src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs

```diff
@@ -560,23 +597,35 @@ private void GetItemsAndContinue(EvalResultDataItem dataItem, DkmWorkList workLi
             var numRows = rows.Count;
             Debug.Assert(count >= numRows);
             var results = new DkmEvaluationResult[numRows];
-            GetEvaluationResultsAndContinue(rows, results, 0, numRows, workList, inspectionContext, value.StackFrame,
-                () => ContinueWithExceptionHandling(
+            var wl = new WorkList(workList, e => completionRoutine(DkmEvaluationEnumAsyncResult.CreateErrorResult(e)));
+            GetEvaluationResultsAndContinue(rows, results, 0, numRows, wl, inspectionContext, value.StackFrame,
+                () => wl.ContinueWith(
                     () =>
                     {
                         completionRoutine(new DkmEvaluationEnumAsyncResult(results));
                         rows.Free();
-                    },
-                    e => completionRoutine(DkmEvaluationEnumAsyncResult.CreateErrorResult(e))));
+                    }));
+            wl.Execute();
         }
 
-        private void StoreResultAndContinue(DkmEvaluationResult result, ArrayBuilder<EvalResultDataItem> rows, DkmEvaluationResult[] results, int index, int numRows, DkmWorkList workList, DkmInspectionContext inspectionContext, DkmStackWalkFrame stackFrame, CompletionRoutine completionRoutine)
+        private void GetEvaluationResultsAndContinue(ArrayBuilder<EvalResultDataItem> rows, DkmEvaluationResult[] results, int index, int numRows, WorkList workList, DkmInspectionContext inspectionContext, DkmStackWalkFrame stackFrame, CompletionRoutine completionRoutine)
         {
-            results[index] = result;
-            index++;
             if (index < numRows)
             {
-                GetEvaluationResultsAndContinue(rows, results, index, numRows, workList, inspectionContext, stackFrame, completionRoutine);
+                CreateEvaluationResultAndContinue(rows[index], workList, inspectionContext, stackFrame,
+                    result => workList.ContinueWith(
+                        () =>
+                        {
+                            results[index] = result;
+                            if (index < numRows - 1)
```

Comment:
Actually, could we remove the if/else from the lambda entirely and just call GetEvaluationResultsAndContinue (relying on the outer if to call the completionRoutine)?